consensus/src/payload_manager.rs

```diff
@@ -348,14 +331,172 @@ impl PayloadManager {
                 block.id()
             ),
         };
+
         if let PayloadManager::InQuorumStore(_, _, Some(consensus_publisher)) = self {
+            let transaction_payload = BlockTransactionPayload::new(
+                transactions.clone(),
+                limit,
+                proof_with_data,
+                inline_batches,
+            );
             let message = ConsensusObserverMessage::new_block_payload_message(
                 block.gen_block_info(HashValue::zero(), 0, None),
-                result.0.clone(),
-                result.1,
+                transaction_payload,
             );
             consensus_publisher.publish_message(message).await;
         }
-        Ok(result)
+
+        Ok((transactions, limit))
+    }
+}
+
+/// Returns the transactions for the consensus observer payload manager
+async fn get_transactions_for_observer(
+    block: &Block,
+    payload: &Payload,
+    block_payloads: &Arc<Mutex<BTreeMap<(u64, Round), BlockPayloadStatus>>>,
+    consensus_publisher: &Option<Arc<ConsensusPublisher>>,
+) -> ExecutorResult<(Vec<SignedTransaction>, Option<u64>)> {
+    // The data should already be available (as consensus observer will only ever
+    // forward a block to the executor once the data has been received and verified).
+    let block_payload = match block_payloads.lock().entry((block.epoch(), block.round())) {
+        Entry::Occupied(mut value) => match value.get_mut() {
+            BlockPayloadStatus::AvailableAndVerified(block_payload) => block_payload.clone(),
+            BlockPayloadStatus::AvailableAndUnverified(_) => {
+                // This shouldn't happen (the payload should already be verified)
+                let error = format!(
+                    "Payload data for block epoch {}, round {} is unverified!",
+                    block.epoch(),
+                    block.round()
+                );
+                return Err(InternalError { error });
+            },
+        },
+        Entry::Vacant(_) => {
+            // This shouldn't happen (the payload should already be present)
+            let error = format!(
+                "Missing payload data for block epoch {}, round {}!",
+                block.epoch(),
+                block.round()
+            );
+            return Err(InternalError { error });
+        },
+    };
+
+    // Verify the payload and inline batches before returning the data. The
```

Comment:
FYI: this (final) verification logic will be moved into consensus observer in the PR (after we clean up the ordered block verification) 😄